EKF3: Sensor Affinity and Accumulated Error Lane Switching

harshitsankhla opened the pull request:
An EKF3 port of the EK2_Affinity PR #11257 with EKF3 accumulated relative core errors based lane switching #14472 

**_For the general audience_** 

**Affinity** - For a vehicle which has multiple instances of Airspeed - Magnetometer - Barometer - GPS, affinity is a way to assign these sensors between separate EKF cores. This is useful when a vehicle is known to have multiple good quality sensors. If a sensor is healthy and can be used for fusion, we want it to be attached to a specific core. If the number of cores is more than the number of sensor instances, then some cores may share the same sensor (specifically the primary one). A simple example for a vehicle which has 2 baros , 2 GPS, 2 airspeeds, 3 mags and 3 IMUs - following would be the allotment if affinity is enabled and all sensors are healthy.

| LANE | 1 | 2 | 3 | 
|------|---|---|---|
|ASD   | 1 | 2 | 1 |
|BAR   | 1 | 2 | 1 |
|GPS   | 1 | 2 | 1 |
|MAG   | 1 | 2 | 3 |

*assuming 1 is primary for all sensors and ignoring the mag switching logic for simplicity.

**Accumulated Error** - The error which we accumulate is representative of how good/bad an EKF core has been over time. This error is relative to the active primary core. A negative relative error for a core means it has been performing better than the active primary, and vice-versa. So now with affinity, we have a 'statistical' way to determine which sensor combination is the best that can be used. For this, the error term (that we accumulate) should be a right indicator of the sensors performance of that core.  The error score now takes into account for innovations from all the above sensors. 

As Tridge pointed out in his affinity PR for EKF2 #11257  - This provides for a more statistically robust way of using multiple sensors when the airframe is known to have multiple "good" sensors. EKF lane changing is used to select the lane with the best sensors.

Some SITL parameters for simulating errors added through this PR -

- Airspeed Offset in m/s
- GPS Velocity in m/s
- Freeze barometer to last reported altitude

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -1021,6 +1024,107 @@ void NavEKF3_core::writeExtNavVelData(const Vector3f &vel, float err, uint32_t t
+        return;
+    }
+
+    // in normal operation use the primary compass
```
this breaks the existing compass switching when affinity is disable for compasses (note that my PR broke that too, I've now fixed it).
When affinity is disabled we want to use the existing compass switching logic which is based on innovations, so just delete this line and leave magSelectIndex as it is

harshitsankhla replied:
fixed here now

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -523,22 +526,22 @@ void NavEKF3_core::readGpsData()
 
-    if (gps.last_message_time_ms() - lastTimeGpsReceived_ms > frontend->sensorIntervalMin_ms) {
-        if (gps.status() >= AP_GPS::GPS_OK_FIX_3D) {
+        if (gps.last_message_time_ms(selected_gps) - lastTimeGpsReceived_ms > 70) {
```
cut&paste bug, in EKF3 we use frontend->sensorIntervalMin_ms

harshitsankhla replied:
fixed, sorry

tridge commented on libraries/AP_Airspeed/AP_Airspeed.cpp:
```diff
@@ -313,6 +313,9 @@ void AP_Airspeed::init()
             delete sensor[i];
             sensor[i] = nullptr;
         }
+        if (sensor[i] != nullptr) {
```
these changes look like copied from my PR, should have been done with cherry-pick to preserve author (and blame!)

harshitsankhla replied:
cherry picked 2 of your commits for baro and airspeed that are used unchanged in this PR. Thanks for the advice, will keep mind of it in future.

tridge commented:
I think the combination of the affinity and the error accumulation looks good. We now need to start developing tests for the ArduPilot autotest suite for the key features in this PR. Specifically:
 - for airspeed failure, such as an airspeed offset error (simulating a pitot failure). This is a really important one, as it affects aircraft flying in rain. We can simulate that in the test suite by changing ARSPD_OFFSET, or adding a SIM_ARSDP_ERR parameter
 - for GPS developing errors. We need to add GPS velocity error parameters to libraries/SITL so we can simulate poor GPS lock where the velocity error increases
 - for mag developing errors, which we can simulate with SIM_MAG offset changes
- for baro developing errors. We'll need a way to add a baro offset per-baro in SITL (and fix AP_Baro to allow for more than one simulated baro)
One tricky thing is detecting lane change in the autotest suite. @peterbarker any recommendations on how to do that?

peterbarker commented:
On Mon, 6 Jul 2020, Andrew Tridgell wrote:

>  *  for airspeed failure, such as an airspeed offset error (simulating a pitot failure). This is a really important one, as it affects aircraft flying in
>     rain. We can simulate that in the test suite by changing ARSPD_OFFSET, or adding a SIM_ARSDP_ERR parameter

See this PR which adds different failure types: 
https://github.com/ArduPilot/ardupilot/pull/13111

>     One tricky thing is detecting lane change in the autotest suite. @peterbarker any recommendations on how to do that?

Could just look for statustext messages - we do emit those, right?

tridge commented:
> Could just look for statustext messages - we do emit those, right?

yes, as long as that is robust enough

harshitsankhla commented:
> * for baro developing errors. We'll need a way to add a baro offset per-baro in SITL (and fix AP_Baro to allow for more than one simulated baro)

I have this up - https://github.com/ArduPilot/ardupilot/pull/14711

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:
```diff
@@ -46,7 +46,7 @@ float NavEKF3_core::errorScore() const
         score = MAX(score, 0.5f * (velTestRatio + posTestRatio));
         // Check altimeter fusion performance
-        score = MAX(score, hgtTestRatio);
+        score = MAX(score, hgtTestRatio*2);
```
experimenting something, missed changing back, sorry

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -523,22 +526,22 @@ void NavEKF3_core::readGpsData()
             // Use the driver specified delay
             float gps_delay_sec = 0;
-            gps.get_lag(gps_delay_sec);
+            gps.get_lag(selected_gps, gps_delay_sec);
```
changed now

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -786,12 +789,12 @@ void NavEKF3_core::readAirSpdData()
+    const auto *airspeed = AP::airspeed();
+    if (airspeed &&
+        airspeed->use(selected_airspeed) &&
+        airspeed->last_update_ms(selected_airspeed) != timeTasReceived_ms) {
```
changed back to the original logic in EKF3, just with the selected_airspeed now

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3.h:
```diff
@@ -584,13 +584,14 @@ class NavEKF3 {
+    uint8_t coreImuIndex[MAX_EKF_CORES];         // IMU index used by this core
+    float coreRelativeErrors[MAX_EKF_CORES];     // relative errors of cores with respect to primary
+    float coreErrorScores[MAX_EKF_CORES];        // the instance error values used to update relative core error
+    uint64_t coreLastTimePrimary[MAX_EKF_CORES]; // last time we were using this core as primary
```
_us here

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -782,6 +782,13 @@ bool NavEKF3::InitialiseFilter(void)
     }
 
+    // set last time the cores were primary to 0
+    if (ret) {
```
memset

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -967,6 +975,7 @@ void NavEKF3::updateCoreRelativeErrors()
 
             // reduce error for a core only if its better than the primary lane by at least the Relative Error Threshold, this should prevent unnecessary lane changes
             if(error > 0 || error < -REL_ERR_THRESH) {
+                //::printf("Reducing error\n");
```
remove this

harshitsankhla commented on libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:
```diff
@@ -47,6 +47,15 @@ float NavEKF3_core::errorScore() const
+        // Check airspeed fusion performance - only when we are using at least 2 airspeed sensors so we can switch lanes with 
+        // a better one. This only comes into effect for a forward flight vehicle. A damping factor of 0.3 is added to keep the
+        // EKF less sensitive to innovations arising due events like strong gusts of wind, thus, prevent reporting high error scores
+        if(assume_zero_sideslip()) {
```
whitespace

harshitsankhla commented:
Test : Simulate a gyroscope failure by sending zero values on all three-axis.
This graph shows a lane switch occurring when the gyroscope of the primary EKF core fails completely mid-flight during mission. Major difference from master is the time taken to make the switch which is nearly 2s faster. Refer to logs below.  

![0Gyro_Graph1](https://user-images.githubusercontent.com/13604433/88758881-2b91ff80-d187-11ea-89b6-729dedf95cea.PNG)
![0Gyro_Graph2](https://user-images.githubusercontent.com/13604433/88758640-8a0aae00-d186-11ea-8856-2d8749edc21b.png)

Logs 
This PR - https://drive.google.com/file/d/1IBv6CRYHvZ9-qBnLUOeBKimI9NpENnqA/view?usp=sharing
Master - https://drive.google.com/file/d/1bN5P7rQbcIu8C0gIn9xi2jtF-v6J9PjM/view?usp=sharing

harshitsankhla commented:
Test (Vehicle:Plane) : Simulate a GPS velocity error by adding an offset in all 3-axis NED.
This graph shows a lane switch occurred on adding a velocity error of 1.5m/s simultaneously along all 3-axis. The new PR is able to switch lanes after the innovation levels increase as we are accumulating error -

![gps_NED_newLS](https://user-images.githubusercontent.com/13604433/88763418-bcb9a400-d190-11ea-8485-f00b0a8dcbb8.png)

whereas there is no switch in master because the instantaneous innovations are not high enough for its logic to switch.

![gps_NED_master](https://user-images.githubusercontent.com/13604433/88763438-c4794880-d190-11ea-94ce-1d007779ee7d.png)

Logs 
This PR - https://drive.google.com/file/d/1KT0wkGmMk1jFOcJR7qml7Ny5kputwHJ7/view?usp=sharing
Master - https://drive.google.com/file/d/1Ph2NOZu7CBe57ox7N214zJkNNDR-jiSo/view?usp=sharing

harshitsankhla commented:
Test (Vehicle: SITL Plane) : Add random noise to the airspeed sensor.
Lane Switching with airspeed affinity enabled and adding airspeed innovations to the EKF errorscore.

![airspeed](https://user-images.githubusercontent.com/13604433/88839963-62086280-d1f9-11ea-9c1c-5334fac7a2f7.png)
![airspeed2](https://user-images.githubusercontent.com/13604433/88839819-31c0c400-d1f9-11ea-9cd4-2fca94817d56.png)

Lane switches due to moderate airspeed noise do not happen in master.

Logs
This PR - https://drive.google.com/file/d/1rXRwBwg8HHqkT4Am5i1euf8NFGMhoaSA/view?usp=sharing
Master - https://drive.google.com/file/d/1jQR92FRcyF9eg0cHbMd-gJBtvWuIvuvd/view?usp=sharing

harshitsankhla commented:
Test(Vehicle: SITL Plane) : Simulate a accelerometer failure by sending zero values on all three-axis.
Similar to the gyro test, this graph shows recovery when the primary accelerometer stops working mid-flight. The time taken to switch is 0.39s faster in this PR than master. Refer to logs below.

![0acc_newLS](https://user-images.githubusercontent.com/13604433/89001033-7e4df180-d316-11ea-9439-f016ea4815d0.png)

Logs
This PR - https://drive.google.com/file/d/1dKeZIJLy0HqmOHZiXk6BMbuxuwsKQWyz/view?usp=sharing
Master - https://drive.google.com/file/d/1e7jzV4t_SUEQWmH0l6ELIKeJhPGw8Pyr/view?usp=sharing

tridge commented:
@harshitsankhla I did a test flight on a small quad with your PR. Log is here:
http://uav.tridgell.net/ekf3-affinity/test1.bin
it switched lanes rapidly while disarmed. I think it is caused by an interaction of the new code and this code:
https://github.com/harshitsankhla/ardupilot/blob/ekf3-affinity/libraries/AP_NavEKF3/AP_NavEKF3.cpp#L891
I also don't see any mag innovation in the error score yet. I thought we discussed adding in mag innovation if mag affinity is enabled? We'd then disable the exiting compass switching

harshitsankhla commented:
Test (Vehicle: SITL Plane) : Add random noise to the barometer. 
This graph shows a lane switch happening when random noise is added to the primary barometer and affinity is enabled. This random noise fluctuates the true height with a random value between -10 to 10m. With the new logic we are able to switch immediately - 
![10BaroG2](https://user-images.githubusercontent.com/13604433/89129911-b7f14900-d51e-11ea-8c05-a5972387c8d1.png)

whereas master does not switch, which is mostly due to the reason that without affinity, there is no better lane it can jump to as both are sharing the same barometer
![Uploading 10BaroG1.png…]()

also with the new logic, the error has to be atleast ~6m for a lane switch to happen, which acts as a threshold to the sensitivity of switching. This was observed in a SITL test by increasing the noise from 0.2 to 7 till a lane switch occurred. Corresponding innovation levels can be seen here -
![6BaroG3](https://user-images.githubusercontent.com/13604433/89129916-c475a180-d51e-11ea-8922-3d899c18f16f.png)

Logs
This PR - https://drive.google.com/file/d/1pt63MhOuKjbranlTpsPnpJoa9_kLiqrE/view?usp=sharing
Master - https://drive.google.com/file/d/1pmGPxuXe2BIwTbIKW-pO3F1RUMK1yczi/view?usp=sharing
This PR (going from 0.2 to 7) - https://drive.google.com/file/d/1_b5su6Rah8Yl3JEPcemCQ8N8omExSPrv/view?usp=sharing

harshitsankhla commented:
Test (Vehicle: SITL Plane) : Change magnetometer offset.
Following graph shows lane switching on changing the magnetometer offset mid-flight. This is observed after adding magTestRatio to the errorscore only when MAG affinity is enabled.
![mag_1](https://user-images.githubusercontent.com/13604433/89641359-23486b80-d8cf-11ea-8217-35017c260ae8.png)
![mag_2](https://user-images.githubusercontent.com/13604433/89641361-23486b80-d8cf-11ea-85f0-a667703972e4.png)

whereas in master, only the compass gets swapped for the primary lane, a short while after sustaining high innovations
![mag3](https://user-images.githubusercontent.com/13604433/89641360-23486b80-d8cf-11ea-8c9f-269515bac3a2.png)

Logs
This PR - https://drive.google.com/file/d/1-nS8yU2EppGhHHCqA_yu_VkfqIcsA9xA/view?usp=sharing
Master - https://drive.google.com/file/d/1SMgkM2nHyJwhKMFLD8TwVOOcs9UYVNl-/view?usp=sharing

harshitsankhla commented:
ping @tridge @priseborough 
can we get any more real flight tests ? it will take at least a week more for me to get my hands on my quad.

tridge commented:
> can we get any more real flight tests ?

yep. My charger had died. I got the new one today so was able to charge and do some testing. Test log here:
http://uav.tridgell.net/ekf3-affinity/test2.bin
This is your branch rebased on master.
I use EK3_AFFINITY=3 for the first flight, and EK3_AFFINITY=0 for the 2nd. It also had EKF2 running.
It is too dark now to fly again today, but I should be able to test mag affinity tomorrow.
Only a small flying area, sorry about the lack of any real speed. I can test in a bigger area another day.

harshitsankhla commented:
Test (Vehicle: SITL Plane) : High random noise on Accelerometer X-Axis
Following graph shows lane switching when the X-Axis of the primary IMU's accelerometer experiences heavy noise. 
![accAliasCompare](https://user-images.githubusercontent.com/13604433/90392976-81ced000-e0ad-11ea-904a-b06ba27f8b11.png)
![alias_x_newls](https://user-images.githubusercontent.com/13604433/90392981-83989380-e0ad-11ea-8618-fb6620d4be54.png)

master is not able to make a lane switch in this test -
![alias_x_master](https://user-images.githubusercontent.com/13604433/90393336-26e9a880-e0ae-11ea-9b47-b2f7775384b9.png)

Logs
This PR - https://drive.google.com/file/d/1Gx0ribAXmFgEelVcvJHRTnGWf8rGskQM/view?usp=sharing
Master - https://drive.google.com/file/d/1DsK6KdzpxSMkdWsuzE6JLHDDr3qtjajg/view?usp=sharing

harshitsankhla commented:
Test (Vehicle: SITL Plane) : Random noise in GPS Velocity on all 3-axis

This log shows a lane switch getting triggered after a velocity error is observed on the GPS on all 3-axis, simultaneously, in the range -2 to +2m.
![gps_velErrRandom2](https://user-images.githubusercontent.com/13604433/90434632-61256b00-e0eb-11ea-9db5-2c27e3e2f42d.png)

master doesn't trigger a lane switch, even after adding extremely high velocity errors, because both lanes are using the same primary GPS, hence both observe high error scores and the system can not find a 'better lane'. (We don't have affinity in master).
![gps_verr_master](https://user-images.githubusercontent.com/13604433/90434744-93cf6380-e0eb-11ea-96de-8177bedf6617.png)

Logs
This PR - https://drive.google.com/file/d/1piF2cjWSjzbtqpB3_RQvgseCqUTiBt5-/view?usp=sharing
Master - https://drive.google.com/file/d/10sOiovUA-IM5PHPrVm-1kJCNECZsgGf3/view?usp=sharing

tridge commented on Tools/autotest/arduplane.py:
```diff
@@ -1765,6 +1765,87 @@ def fly_terrain_mission(self):
+
+            self.progress("Gyroscope: Applying offset to 2nd Gyroscope")
+            self.context_push()
+            self.wait_statustext(text="lane switch", timeout=10, the_function=self.set_parameter("INS_GYR2OFFS_Y", 1))
```
we need to parse these in a more detailed fashion, so we check it does the correct lane change, not just that it changed lane at all.

khancyr replied:
I don't really know what you are logging, but you should be able to use something similar to https://github.com/ArduPilot/ardupilot/blob/master/Tools/autotest/common.py#L5212 to look into the current log and search for the right message!

harshitsankhla replied:
parsing the string seems fine now, we can check the XKF4.PI (EKF3 Primary Core Index) but probably would also have to cross check with the time-stamp and might over complicate things. Will look into this option if the string parsing fails anywhere.

tridge commented:
> Test (Vehicle: SITL Plane) : Random noise in GPS Velocity on all 3-axis

I expected deeper analysis than this. For a start, plotting the error score and looking at whether it behaved as expected. I also see no testing of the magnetometer affinity, and the autotest is very weak, not even checking that it changes to the correct lane

tridge commented on libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:
```diff
@@ -2156,6 +2156,17 @@ bool AP_AHRS_NavEKF::have_ekf_logging(void) const
+int8_t AP_AHRS_NavEKF::get_active_airspeed_index() const
+{
+// we only have affinity for EKF3 as of now
+#if HAL_NAVEKF3_AVAILABLE
```
this needs to check if EKF3 is the active EKF instance (EKF3 may be compiled in but not being used). You can use active_EKF_type() to determine which EKF is active

harshitsankhla replied:
fixed, thanks

tridge commented:
> In this test, a random noise of +-20m/s is added to the primary lane's airspeed sensor.

20 m/s random noise isn't really realistic for airspeed errors. Here is a graph of a fairly typical airspeed innovation:
![image](https://user-images.githubusercontent.com/831867/90726219-b38da400-e304-11ea-93ca-bfc1e6925869.png)

tridge commented:
needs a rebase

tridge commented:
Here is a log demonstrating airspeed affinity saving the aircraft in SITL with simulated pitot blockage:
http://uav.tridgell.net/EKF/Affinity/airspeed-block-test1.bin

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:
```diff
@@ -45,6 +45,20 @@ float NavEKF3_core::errorScore() const
+        // EKF less sensitive to innovations arising due events like strong gusts of wind, thus, prevent reporting high error scores
+        if (assume_zero_sideslip()) {
+            const auto *arsp = AP::airspeed();
+            if (arsp->get_num_sensors() >= 2 && frontend->_affinity & EKF_AFFINITY_ARSP) {
```
use () around bitops

tridge replied:
we may want to scale the airspeed contribution with 1/sqrt(airspeed) to prevent too much contribution at low speeds?

tridge commented on libraries/AP_AHRS/AP_AHRS.h:
```diff
@@ -287,6 +287,12 @@ class AP_AHRS
 
+    // return true if airspeed comes from a specific airspeed sensor, as
+    // opposed to an IMU estimate
+    bool airspeed_sensor_enabled(int8_t airspeed_index) const {
```
add check for airspeed_index < 0

tridge commented on libraries/AP_AHRS/AP_AHRS_NavEKF.h:
```diff
@@ -286,6 +286,13 @@ class AP_AHRS_NavEKF : public AP_AHRS_DCM {
+    // over to a lane not using the primary airspeed sensor, so AHRS should know which airspeed sensor
+    // to use, i.e, the one being used by the primary lane. A lane switch could have happened due to an 
+    // airspeed sensor fault, which makes this even more necessary
+    int8_t get_active_airspeed_index() const;
```
should this be unsigned?

tridge commented:
> this log here shows for noise <~6m in the primary barometer does not cause a change in the relative error score for other lanes.

a better test would be an offset, or a fixed baro pressure

khancyr commented on Tools/autotest/arduplane.py:
```diff
@@ -1765,6 +1765,109 @@ def fly_terrain_mission(self):
+
+        self.install_message_hook(statustext_hook)
+        
+        self.context_push()
```
`self.context_push()` should be at the start of the test before starting to modify the parameter: it say : starting from there I save the parameter modification. That will allow to have `self.context_pop()` at the end of the test that rewind the parameter change and let the state of SITL clean for further test!

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -639,6 +639,15 @@ const AP_Param::GroupInfo NavEKF3::var_info[] = {
+    // @Description: These options control the affinity between sensor instances and EKF cores
+    // @User: Advanced
+    // @Bitmask: 0:EnableGPSAffinity,1:EnableBaroAffinity,2:EnableCompassAffinity,3:EnableAirspeedAffinity
+    AP_GROUPINFO("AFFINITY", 61, NavEKF3, _affinity, 0),
```
Unless we have tested the ability to cope with changes to these settings without rebooting, then I would recommend labelling this as

// @RebootRequired: True

harshitsankhla replied:
i don't think a reboot is required, because the sensor selection functions which assign the sensors to respective cores first assign the primary instance and then change the assignment only if affinity for that respective sensor is available. This happens whenever we update the filter. So if affinity is turned off mid-flight after being on (or vice-versa) we would still be using the primary instance of that sensor for all cores (or affinity based assignment). Maybe @tridge can better judge this ?

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:
```diff
@@ -45,6 +45,20 @@ float NavEKF3_core::errorScore() const
         // Check altimeter fusion performance
         score = MAX(score, hgtTestRatio);
+        // Check airspeed fusion performance - only when we are using at least 2 airspeed sensors so we can switch lanes with 
+        // a better one. This only comes into effect for a forward flight vehicle. A damping factor of 0.3 is added to keep the
```
Suggest replacing 'damping' with 'sensitivity'

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -1024,6 +1027,104 @@ void NavEKF3_core::writeExtNavVelData(const Vector3f &vel, float err, uint32_t t
+            // use core_index airspeed if it is healthy
+            selected_airspeed = core_index;
+        }
+    }
```
suggest moving the non affinity case as an else condition here to avoid unnecessary airspeed object access when doing airspeed affinity, eg:

    if (frontend->_affinity & EKF_AFFINITY_ARSP &&
        core_index < arsp->get_num_sensors() &&
        arsp->healthy(core_index) &&
        arsp->use(core_index))
    {
        // use core_index airspeed if it is healthy
        selected_airspeed = core_index;
    } else {
        // in normal operation use the primary airspeed sensor
        selected_airspeed = arsp->get_primary();
    }

priseborough commented on libraries/SITL/SITL.h:
```diff
@@ -152,12 +152,15 @@ class SITL {
     AP_Float arspd_fail_pitot_pressure; // 1st pitot tube failure pressure
     AP_Float arspd2_fail_pressure; // 2nd pitot tube failure pressure
     AP_Float arspd2_fail_pitot_pressure; // 2nd pitot tube failure pressure
+    AP_Float arspd_offset; // 1st airspeed sensor offset
```
// 1st airspeed sensor offset in m/s

priseborough commented on libraries/SITL/SITL.h:
```diff
@@ -152,12 +152,15 @@ class SITL {
     AP_Float arspd2_fail;   // 2nd pitot tube failure
     AP_Float arspd_fail_pressure; // 1st pitot tube failure pressure
     AP_Float arspd_fail_pitot_pressure; // 1st pitot tube failure pressure
     AP_Float arspd2_fail_pressure; // 2nd pitot tube failure pressure
```
// 1st pitot tube failure pressure in Pa

tridge commented on Tools/autotest/arduplane.py:
```diff
@@ -1834,6 +1834,167 @@ def fly_terrain_mission(self):
+            # Cleanup
+            self.set_rc(2, 1500)
+            self.set_parameter("SIM_BARO2_FREEZE", old_parameter)
+            self.lane_switches = []
```
I know I suggested this, but I realise now this introduces a race condition by clearing the list of switches. If we got two lane switches in one test then we would pass when we should fail.
We should remove the "self.lane_switches = []" after each test, and instead check for the accumulated set of lane switches, so the 2nd test would be:
 if self.lane_switches = [1,0]
then next one would be
 if self.lane_switches = [1,0,1]
etc

harshitsankhla replied:
I see, I'll make that change and test it.

harshitsankhla commented:
Listing a few things still needed to be decided - 

1. As Tridge mentioned, scaling the airspeed innovations with 1/sqrt(airspeed)  here - https://github.com/ArduPilot/ardupilot/pull/14674/commits/1e3388e4a0920ab71204d8defa28e3ef107179d8#diff-4631b96a5887525e54d1f1cebe8fc368R54

2. Keep/Remove the state predict enabled check - https://github.com/ArduPilot/ardupilot/pull/14674/commits/1e3388e4a0920ab71204d8defa28e3ef107179d8#diff-373e2bd2962de50bed30c0eed4156258L825

3. Added this for avoiding a segmentation fault with copter. Is it OK ? - https://github.com/ArduPilot/ardupilot/pull/14674/commits/c4fd278c09090b516845ae68ad56988ce625223f#diff-34e6170614ab781f6d2cba61f0ae71b6R548

4. Changed the 0.2 error threshold to a parameter. Also should we add some protection for it not being set to, say, a negative value ? It seems a little risky. - https://github.com/ArduPilot/ardupilot/pull/14674/commits/1e3388e4a0920ab71204d8defa28e3ef107179d8#diff-373e2bd2962de50bed30c0eed4156258R648

tridge commented:
>     1. As Tridge mentioned, scaling the airspeed innovations with 1/sqrt(airspeed)  here

I discussed this a bit with Paul, and we don't think we should change this just yet, but will instead wait for some test results on real aircraft with dual airspeed sensors.

> 2. Keep/Remove the state predict enabled check

I think this should be removed as it could force us to move to an inferior lane

> 3. Added this for avoiding a segmentation fault with copter. Is it OK ?

I'd like to see the backtrace of this, or know how to reproduce

> 4. Changed the 0.2 error threshold to a parameter. 

looks good, except in the param doc comments I think the increment should be smaller, probably 0.05, and range start at 0.05

> Also should we add some protection for it not being set to, say, a negative value 

using MAX(value, 0.01) would be reasonable.

tridge commented:
> Added this for avoiding a segmentation fault with copter. Is it OK ? 

I've reproduced it, and I think it would be better to make get_active_airspeed_index() return 0 if _airspeed is nullptr

tridge commented:
I've fixed up some code style issues and removed the duplicate airspeed_estimate() implementation, and force pushed

tridge commented:
this is ready to merge when it passes CI

rmackay9 commented on libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:
```diff
@@ -543,8 +543,8 @@ Vector3f AP_AHRS_NavEKF::wind_estimate(void) const
 bool AP_AHRS_NavEKF::airspeed_estimate(float &airspeed_ret) const
-{
-    return AP_AHRS_DCM::airspeed_estimate(airspeed_ret);
+{    
```
Not a blocker but a small white space issue here..

harshitsankhla replied:
fixed, thanks

rmackay9 commented on libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:
```diff
@@ -2160,6 +2160,22 @@ bool AP_AHRS_NavEKF::have_ekf_logging(void) const
 }
 
+//get the index of the active airspeed sensor, wrt the primary core
+uint8_t AP_AHRS_NavEKF::get_active_airspeed_index() const
```
normally we put the bracket at the end of the line above in the .cpp files.

rmackay9 commented on libraries/AP_Baro/AP_Baro_SITL.cpp:
```diff
@@ -76,6 +76,15 @@ void AP_Baro_SITL::_timer()
+        // if freezed barometer, report altitude to last recorded altitude
+        if (_sitl->baro_freeze[_instance] == 1) {
+            sim_alt = _last_altitude;
+        }
```
small formatting issue here, the else should be on the same line as the bracket before it

harshitsankhla replied:
fixed, thanks

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -814,37 +839,64 @@ void NavEKF3::UpdateFilter(void)
+
+        bool betterCore = false;
+        bool altCoreAvailable = false;
+        float maxCoreError = 0; // looking for cores that have error lower than the current primary
```
is it a maxCoreError or a minCoreError?

harshitsankhla replied:
yes, indeed makes more sense

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -906,6 +958,45 @@ void NavEKF3::requestYawReset(void)
+{
+    float error = 0;
+    for (uint8_t i = 0; i < num_cores; i++) {
+        if(i != primary) {
```
NitPick: should have a space after the "if"

harshitsankhla replied:
fixed, thanks

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -906,6 +958,45 @@ void NavEKF3::requestYawReset(void)
+        if(i != primary) {
+            error = coreErrorScores[i] - coreErrorScores[primary];
+            // reduce error for a core only if its better than the primary lane by at least the Relative Error Threshold, this should prevent unnecessary lane changes
+            if(error > 0 || error < -MAX(_err_thresh, 0.05)) {
```
same nitpick here, should have a space after the "if".

harshitsankhla replied:
fixed, thanks

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -1119,6 +1210,39 @@ uint8_t NavEKF3::getActiveMag(int8_t instance) const
+uint8_t NavEKF3::getActiveBaro(int8_t instance) const
+{
+    if (instance < 0 || instance >= num_cores) instance = primary;
+    if (core) {
```
should this be "if (core[instance]) {?

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -523,22 +526,22 @@ void NavEKF3_core::readGpsData()
 
-    if (gps.last_message_time_ms() - lastTimeGpsReceived_ms > frontend->sensorIntervalMin_ms) {
-        if (gps.status() >= AP_GPS::GPS_OK_FIX_3D) {
+        if (gps.last_message_time_ms(selected_gps) - lastTimeGpsReceived_ms > frontend->sensorIntervalMin_ms) {
```
I guess the "selected" gps is the same as the "active" gps?  It might be nice to keep the number of synonyms to a minimum because it makes the developer stop for a moment to wonder if they are the same or not.  i.e. It might be good to rename "selected_gps" to "active_gps".

harshitsankhla replied:
we might want to keep selected_gps so it is consistent with the other 2 (selected_baro, and selected_airspeed)

rmackay9 commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -1028,6 +1031,104 @@ void NavEKF3_core::writeExtNavVelData(const Vector3f &vel, float err, uint32_t t
+    auto &gps = AP::gps();
+
+    // in normal operation use the primary GPS
+    selected_gps = gps.primary_sensor();
```
"active", "selected", "preferred" and "primary".. that's a lot of words to keep track of..

harshitsankhla replied:
indeed, we can do a separate 'cleansing' PR later

rmackay9 commented:
I haven't read the description and comments above but I wonder about two things in particular:

1. how does this work with GPS blending?  If it doesn't work we should probably add a check somewhere to ensure that blending and affinity are not turned on at the same time
2. for naming, I think we should try and get rid of at least one of these three synonyms, "active", "selected", "preferred".  It looks to me like "active" is the least used and always means exactly "selected".

tridge commented:
>     1. how does this work with GPS blending?  If it doesn't work we should probably add a check somewhere to ensure that blending and affinity are not turned on at the same time

I presume you are referring to the GPS affinity option? It does work fine to use that in combination with blending. The first lane would use first GPS, 2nd lane would use 2nd GPS and 3rd lane would use the blended GPS. This is because we define the blended GPS as GPS index 2 in AP_GPS

>     2. for naming, I think we should try and get rid of at least one of these three synonyms, "active", "selected", "preferred".  It looks to me like "active" is the least used and always means exactly "selected".

there are 143 uses of 'active' in EK3 at the moment, 13 of 'selected', but regardless, changing to use one of these words would require significant changes to code other than what this PR is about. I don't think it is in scope for this PR

harshitsankhla commented:
forced push with the formatting corrections suggested above
